examples/image_captioning/model.py

```diff
@@ -0,0 +1,267 @@
+        reporter.report({'loss': loss}, self)
+
+        return loss
+
+    def predict(self, imgs, bos, eos, max_caption_length):
+        """Batch of images to captions."""
+        imgs = Variable(imgs)
+        img_feats = self.feat_extractor(imgs)
+        captions = self.lang_model.predict(
+            img_feats, bos=bos, eos=eos, max_caption_length=max_caption_length)
+        return captions
+
+
+class VGG16FeatureExtractor(chainer.Chain):
+
+    """VGG16 image feature extractor."""
+
+    def __init__(self):
+        super(VGG16FeatureExtractor, self).__init__()
+        with self.init_scope():
+            self.cnn = L.VGG16Layers()
+        self.cnn_layer_name = 'fc7'
+
+    def prepare(self, img):
+        """Single image to resized and normalized image."""
+        return L.model.vision.vgg.prepare(img)
+
+    def __call__(self, imgs):
+        """Batch of images to image features."""
+        h = self.cnn(imgs, [self.cnn_layer_name])[self.cnn_layer_name]
+        img_feat = F.dropout(F.relu(h), ratio=0.5)
+        return img_feat
+
+
+class LSTMLanguageModel(chainer.Chain):
+
+    """Recurrent LSTM language model.
+
+    Generate captions given features extracted from images.
+    """
+
+    def __init__(self, vocab_size, hidden_size, dropout_ratio, ignore_label):
+        super(LSTMLanguageModel, self).__init__()
+        with self.init_scope():
+            self.embed_word = L.EmbedID(
+                vocab_size,
+                hidden_size,
+                initialW=initializers.Normal(1.0),
+                ignore_label=ignore_label
+            )
+            self.embed_img = L.Linear(
+                hidden_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
+            )
+            self.lstm = L.LSTM(hidden_size, hidden_size)
+            self.out_word = L.Linear(
+                hidden_size,
+                vocab_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
+            )
+
+        self.dropout_ratio = dropout_ratio
+
+    def __call__(self, img_feats, captions):
+        """Batch of image features and image captions to a singe loss.
+
+        Compute the softmax cross-entropy captioning loss.
+        """
+        self.reset(img_feats)
+
+        loss = 0
+        size = 0
+        caption_length = captions.shape[1]
+        for i in range(caption_length - 1):
+            # Compute the loss based on the prediction of the next token in the
+            # sequence
+            x = Variable(self.xp.asarray(captions[:, i]))
+            t = Variable(self.xp.asarray(captions[:, i + 1]))
+            if (t.data == self.embed_word.ignore_label).all():
+                # The next tokens in the sequence should not be accounted for
+                # in the loss, e.g. the targets are paddings that were used to
+                # align the lengths of all captions within the same batch
```

Comment:
Could you clarify these comment?
e.g.
Preprocessed captions are padded to reach a maximum length. So, if all targets are pads, prediction is no longer required.

Reply:
Will fix.